File: _API_TESTING_GHERKIN_/DEMOAPP003_TYPESCRIPT_PLAYWRIGHT/screenplay/tasks/SendGetRequest.ts
import type { APIResponse } from "@playwright/test";
import { CallAnApi } from "../abilities/CallAnApi";
import type { Task } from "../support/types";
import type { Actor } from "../actors/Actor";

export class SendGetRequest implements Task {
  private constructor(private readonly path: string, private readonly params?: Record<string, string>) {}

  static to(path: string, params?: Record<string, string>): SendGetRequest {
    return new SendGetRequest(path, params);
  }

  async performAs(actor: Actor): Promise<void> {
    const api = actor.abilityTo(CallAnApi);
    const response = await api.get(this.path, this.params);
    actor.remember<APIResponse>("last-response", response);
  }
}

File: _API_TESTING_GHERKIN_/DEMOAPP004_PYTHON_PLAYWRIGHT/screenplay/tasks/send_get_request.py
"""Task that performs a GET request via Playwright."""

from __future__ import annotations

from typing import Any, Dict

from screenplay.abilities.call_an_api import CallAnApi
from screenplay.support.memory import ActorMemory
from screenplay.support.memory_keys import MemoryKeys


class SendGetRequest:
    def __init__(self, endpoint: str, params: Dict[str, Any] | None = None):
        self.endpoint = endpoint
        self.params = params or {}

    def perform_as(self, actor):
        """Execute the HTTP request via Playwright."""
        api = actor.ability(CallAnApi)
        response = api.context.get(self.endpoint, params=self.params)
        actor.memory.remember(MemoryKeys.LAST_RESPONSE, response)
